src/Network/Riak/Cluster.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveDataTypeable    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Network.Riak.Cluster
+    ( Cluster(..)
+    , InClusterError(..)
+    , connectToCluster
+    , inCluster
+    , Riak.create
+    , Riak.defaultClient
+    ) where
+
+import           Control.Exception
+import           Control.Exception.Enclosed
+import           Control.Monad.Catch           (MonadThrow (..))
+import           Control.Monad.Trans.Control   (MonadBaseControl)
+import           Data.Typeable
+import           Data.Vector                   (Vector)
+import qualified Data.Vector                   as V
+import           Network.Riak                  (Connection)
+import qualified Network.Riak                  as Riak
+import qualified Network.Riak.Connection.Pool  as Riak
+import           System.Random.Mersenne.Pure64
+import           System.Random.Shuffle         (shuffle')
+
+-- | Datatype holding connection-pool with all known cluster nodes
+data Cluster = Cluster
+    { clusterPools :: Vector Riak.Pool
+      -- ^ Vector of connection pools to riak cluster nodes
+    , clusterGen   :: PureMT
+    }
+    deriving (Show)
+
+-- | Error that gets thrown whenever operation couldn't succeed with
+-- any node.
+data InClusterError = InClusterError [SomeException]
+    deriving (Show, Typeable)
+instance Exception InClusterError
+
+-- | Function to connect to riak cluster with sane pool defaults
+connectToCluster :: [Riak.Client] -> IO Cluster
+connectToCluster clients = do
+    pools <- mapM (\c -> Riak.create c 1 10 20) clients
+    connectToClusterWithPools pools
+
+-- | Function to connect to riak cluster with pre-created list of
+-- 'Riak.Pool' objects
+connectToClusterWithPools :: [Riak.Pool] -> IO Cluster
+connectToClusterWithPools pools = do
+    mt <- newPureMT
+    return (Cluster (V.fromList pools) mt)
+
+-- | Tries to run some operation for a random riak node. If it fails,
+-- tries all other nodes. If all other nodes fail - throws
+-- 'InClusterError' exception.
+inCluster :: (MonadThrow m, MonadBaseControl IO m)
+          => Cluster -> (Connection -> m a) -> m a
+inCluster rc f = do
+    let pools = shuffle' (V.toList (clusterPools rc))
+                         (V.length (clusterPools rc))
+                         (clusterGen rc)
+    go pools []
+  where
+    go [] errors = throwM (InClusterError errors)
+    go (p:ps) es = Riak.withConnectionM p $ \c -> do
+        er <- tryAny (f c)
+        either (\err -> go ps (err:es))
+               return er
```

Comment:
I like the idea of providing this type of function, looking at say the [ruby client](https://github.com/basho/riak-ruby-client/blob/master/lib/riak/client.rb#L341-L376) it does provide this sort of abstraction over a cluster.

However I'd prefer to have the retry behaviour as something that could be passed into `inCluster` and provide a sensible default retry policy that can be plugged in. That would allow for default sensible behaviour and also allow someone who knows what they're doing to customise the retry behaviour. 

```haskell
inCluster :: (MonadThrow m, MonadBaseControl IO m) => 
             RetryPolicy -> Cluster -> (Connection -> m a) -> m a

defaultRetryPolicy :: RetryPolicy
```

We're using `Control.Retry` to do something similar, there's a few examples here https://github.com/ambiata/mismi/blob/59aca51692f007d9561dfb9b515d28940cb0813b/mismi-core/src/Mismi/Environment.hs